uber/server.py

```diff
@@ -135,6 +137,17 @@ def error_page_404(status, message, traceback, version):
 static_overrides(join(c.MODULE_ROOT, 'static'))
 
 
+jsonrpc_services = {}
+
+def register_jsonrpc(service, name=None):
+    name = name or service.__name__
+    assert name not in jsonrpc_services, '{} has already been registered'.format(name)
+    jsonrpc_services[name] = service
+
+jsonrpc_handler = _make_jsonrpc_handler(jsonrpc_services, precall=jsonrpc_reset)
+cherrypy.tree.mount(jsonrpc_handler, join(c.PATH, 'jsonrpc'), c.APPCONF)
```

Comment:
Is there any reason to expose a different path for each service?  We're already namespacing the different "modules", e.g. we call ``attendee.lookup`` or ``shifts.lookup``, etc.  And the APIToken stuff means that we don't need to set different authenticators for the different endpoints.  Is there any reason then to not just expose them all under the same endpoint?

Along those same lines, if you register a service normally via ``services.register(AttendeeLookup(), 'attendee')`` then it's automatically exposed over both JSON-RPC and websocket RPC, though the latter wouldn't matter since our Nginx config is not set up to reverse proxy websocket traffic.  I'm guessing you made this jsonrpc only since the websocket stuff wouldn't work?

One note for our future selves is that Sideboard actually has a ``ws.header_fields`` config settings at https://github.com/magfest/sideboard/blob/master/sideboard/configspec.ini#L61 which is specifically for situations like this.  We could specify that the ``X-Auth-Token`` header is made available via the ``threadlocal`` utility class so that our RPC handler methods can still access its value.  That allows us to write authentication code which will work for both HTTP and websocket handlers.  You can do the same thing with session values via the ``ws.session_fields`` config setting in Sideboard as well.

No need to add that now, since we don't use websockets at the moment, but I wanted to make a note of it, since Sideboard totally supports this kind of authentication across both HTTP and websocket requests.

Reply:
I can't imagine why *.pyc was never added to our `.gitignore`, but I will remove the *.pyc files and update our `.gitignore`.

Regarding your code comments, this **does not** expose each service using a different path. I'm not sure what would lead you to believe it would, but they are all exposed via `https://SERVER/uber/jsonrpc`.

Second, the sideboard service registration facility is terrible. It provides no way to use a custom path to expose services. All services will be exposed under `/jsonrpc` under the root server URL. Our web UI is exposed under `/uber`, but there's no way to expose our services under `/uber/jsonrpc`. And god forbid we ever want to use a different path than `/jsonrpc`. Want to expose your API under `/api`? Nope! You must use `/jsonrpc` or nothing at all!

Also, the authentication facilities provided by sideboard are terrible. I would have loved to be able to use the [auth registry](https://github.com/magfest/sideboard/blob/22295ecc6943d39fc74ebe016ba06a37115709ac/sideboard/lib/_cp.py#L220), but it's almost completely useless. You are only allowed to register a `check()` function that takes no arguments. You have no access to the context in which the function is called. Perhaps it's because I'm attempting to perform both _authentication_ and _authorization_ in the same step. I can quite easily extract the auth token and authenticate it while in the `check()` function, but I have no way of determining if the token has the required access levels (_authorization_) because the `check()` function is not given the request context. Sure, I could examine the request and try to parse out the service and method, and then look up the service object/method and query it to determine the required access levels. But then we'd have two different systems doing routing and dispatch! Madness!

Furthermore, the auth registry assumes you're using the web interface! Why doesn't it just do auth? Why is the authentication code trying to do HTTPRedirects?

I don't know why we wouldn't use WSGI (or some similar scheme) to pass the current context through a chain of middleware, but whatever. I am not a fan of how we do authentication and authorization, and I feel the facilities provided by sideboard are unusable in their current state.

Reply:
Responding slightly out of order.

> Regarding your code comments, this does not expose each service using a different path. I'm not sure what would lead you to believe it would, but they are all exposed via https://SERVER/uber/jsonrpc.

Huh - I also have no idea what would have led me to believe that.  I guess I have to plead "it was after 1am and I must have been more tired than I thought" because I totally remember ``cherrypy.tree,mount`` being called multiple times within a loop, once for each service being exposed.  Like I somehow read that ``join`` as constructing URLs like ``/uber/attendee_lookup/jsonrpc``, which really couldn't be further from what's happening.

I sometimes misread code but not usually THAT badly - weird!

> @EliAndrewC my apologies for the extra salty reply. I had just finished a long day poring over the sideboard code when you chimed in with "Hey why don't you use these tools that have been frustrating you all day?!?!" I was tired and irritable, and I took it out on you. I am sorry.

Apology accepted :heart: 

I should probably clarify what I was suggesting though, since I clearly didn't communicate it well enough, and it would be useful to be on the same page about how we want Reggie/Sideboard development and contributions to work going forward.

The Sideboard authentication facilities as they exist now are EXTREMELY bare and were basically the most minimal thing that was still useful.  It solved a problem I had at work at the time, but it punts on a number of extremely important things, such as not caring about the difference between authentication and authorization, and not being designed to work for more than just vanilla HTTP, etc.

I never intended to suggest that you should have used more of the utilities that Sideboard provides to implement this feature.  What I was trying to say with my "note for our future selves" was that in the future we could use some of what Sideboard already has to do a combination of the following:
- Expose these methods over both WebSocket RPC and JSON-RPC in the future at whatever point we may start to care about websockets.
- In the process, fold some of what we're doing here back into Sideboard and/or further extend Sideboard as necessary to better support this kind of thing.

I understand that this isn't going to be a priority for MAGFest anytime soon, and it may be worth having a conversation about whether this kind of thing will EVER be something that MAGFest does.  Although originally I'd hoped that Sideboard would continue to be used both by MAGFest and at my day job with each organization contributing changes back to the open source project, the direction we take for Reggie may cause this to no longer make sense for MAGFest.  If that is the case or if it becomes the case then comments like this are probably "worthles" since they assume a level of investment in Sideboard which doesn't actually exist.

>I don't know why we wouldn't use WSGI (or some similar scheme) to pass the current context through a chain of middleware

Assuming that we DO care about how Sideboard is designed w.r.t. this kind of thing, the direction Sideboard has been moving for this kind of thing is to use the ``threadlocal`` utility to provide this kind of context.  The intention is that we have four main contexts in which code is run:
1) HTTP request handlers
2) WebSocket RPC
3) Background daemon tasks
4) command line invocations made not as part of a running service

Originally we basically just said, "Ok, I guess (1) should reference cherrypy.request and cherrypy.session and such, and everything else just won't."

More recently we've started moving towards saying, "We should throw everything in ``threadlocal`` since that's always available.  I mentioned ``ws.header_fields`` and ``ws.session_fields`` as examples of this, but certainly didn't mean to say that they were things you should be using now, both since we aren't using websockets and since Sideboard itself has only started moving in this direction and has not yet done anything like standardize which ``threadlocal`` fields should always exist, etc.

None of that is to say that this is definitely the One Right Sideboard Way, I just figured it was worth explaining the approach that we're currently taking and the direction we've been moving with these kinds of features at my day job.